boa_engine/src/builtins/iterable/async_from_sync_iterator.rs

```diff
@@ -0,0 +1,435 @@
+
+        let result = match (throw, args.get(0)) {
+            // 7. If throw is undefined, then
+            (None, _) => {
+                // a. Perform ! Call(promiseCapability.[[Reject]], undefined, « value »).
+                promise_capability
+                    .reject()
+                    .call(
+                        &JsValue::Undefined,
+                        &[args.get_or_undefined(0).clone()],
+                        context,
+                    )
+                    .expect("cannot fail according to spec");
+
+                // b. Return promiseCapability.[[Promise]].
+                return Ok(promise_capability.promise().clone().into());
+            }
+            // 8. If value is present, then
+            (Some(throw), Some(value)) => {
+                // a. Let result be Completion(Call(throw, syncIterator, « value »)).
+                throw.call(&sync_iterator.clone().into(), &[value.clone()], context)
+            }
+            // 9. Else,
+            (Some(throw), None) => {
+                // a. Let result be Completion(Call(throw, syncIterator)).
+                throw.call(&sync_iterator.clone().into(), &[], context)
+            }
+        };
+
+        // 10. IfAbruptRejectPromise(result, promiseCapability).
+        if_abrupt_reject_promise!(result, promise_capability, context);
+
+        // 11. If Type(result) is not Object, then
+        let result = if let Some(result) = result.as_object() {
+            result
+        } else {
+            // a. Perform ! Call(promiseCapability.[[Reject]], undefined, « a newly created TypeError object »).
+            promise_capability
+                .reject()
+                .call(
+                    &JsValue::Undefined,
+                    &[context.construct_type_error("iterator throw function returned non-object")],
+                    context,
+                )
+                .expect("cannot fail according to spec");
+
+            // b. Return promiseCapability.[[Promise]].
+            return Ok(promise_capability.promise().clone().into());
+        };
+
+        // 12. Return AsyncFromSyncIteratorContinuation(result, promiseCapability).
+        Self::continuation(
+            &IteratorResult {
+                object: result.clone(),
+            },
+            &promise_capability,
+            context,
+        )
+    }
+
+    /// `AsyncFromSyncIteratorContinuation ( result, promiseCapability )`
+    ///
+    /// More information:
+    ///  - [ECMA reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-asyncfromsynciteratorcontinuation
+    fn continuation(
+        result: &IteratorResult,
+        promise_capability: &PromiseCapability,
+        context: &mut Context,
+    ) -> JsResult<JsValue> {
+        // 1. NOTE: Because promiseCapability is derived from the intrinsic %Promise%, the calls to promiseCapability.[[Reject]] entailed by the use IfAbruptRejectPromise below are guaranteed not to throw.
```

Comment:
I would try to make these comments less than 100 characters wide, as per Rust formatting guidelines.